modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java

```diff
@@ -2644,78 +2651,128 @@ private synchronized void stop0(boolean cancel, boolean waitForBackups) {
-                                if (log != null) {
-                                    LT.warn(log, "Ignoring potential data loss on cache without backups [name=" +
-                                        cache.getName() + "]");
+                        if (grpCtx.config().getCacheMode() == PARTITIONED && grpCtx.config().getBackups() == 0) {
+                            U.warn(log, "Ignoring potential data loss on cache without backups [name="
+                                + grpCtx.cacheOrGroupName() + "]");
+
+                            continue;
+                        }
+
+                        GridDhtPartitionFullMap fullMap = grpCtx.topology().partitionMap(false);
+
+                        int cacheSpecificAmountOfOwners = -1;
+
+                        if (fullMap != null) {
+                            nodesToExclude.retainAll(fullMap.keySet());
+
+                            int parts = grpCtx.topology().partitions();
+
+                            for (int part = 0; part < parts; part++) {
+                                Set<UUID> partIdealAssignment = grpCtx.affinity().idealAssignmentRaw().get(part).
+                                    stream().map(ClusterNode::id).collect(Collectors.toSet());
+
+                                int cnt = 0;
+
+                                for (Map.Entry<UUID, GridDhtPartitionMap> entry : fullMap.entrySet()) {
+                                    if (nodesToExclude.contains(entry.getKey()))
+                                        continue;
+
+                                    // Skip all non-ideal assignments.
+                                    if (!partIdealAssignment.contains(entry.getKey()))
+                                        continue;
+
+                                    if (entry.getValue().get(part) == GridDhtPartitionState.OWNING)
+                                        cnt++;
                                 }
 
-                                continue;
-                            }
+                                if (cacheSpecificAmountOfOwners == -1)
+                                    cacheSpecificAmountOfOwners = cnt;
 
-                            if (log != null) {
-                                LT.warn(log, "Delaying node shutdown to avoid data loss on cache [name=" +
-                                    cache.getName() + ", grp=" + cctx.group().name() + "]");
+                                cacheSpecificAmountOfOwners = Math.min(cacheSpecificAmountOfOwners, cnt);
                             }
+                        }
+
+                        if (cacheSpecificAmountOfOwners <= 1) {
+                            safeToStop = false;
 
-                            blocked = true;
+                            break;
                         }
+
                     }
 
-                    if (blocked) {
+                    safeToStop = safeToStop && topVer == grid.context().discovery().topologyVersion();
```

Comment:
What's the point for this check?
We can't protect from parition loss if safe shutdown is performed simultaneously with sudden node leave anyway.

Reply:
Reworked, using comparison readyTopologyVersion of each cache group.